File: frontend/src/types/portfolio.ts
export type ReportingCurrency = "CAD" | "USD";

export interface PortfolioSummary {
  as_of_date: string | null;
  reporting_currency: string;
  total_market_value: number;
  holding_count: number;
  account_count: number;
  institution_count: number;
  message?: string;
}

export interface HoldingRow {
  account_id: number;
  account_name: string;
  account_type: string;
  institution: string;
  instrument_id: number;
  symbol: string | null;
  instrument_name: string;
  asset_class: string;
  currency: string;
  quantity: number;
  market_price: number | null;
  market_value: number;
  as_of_date: string;
  sector?: string;
  tags?: string[];
}

export interface ExposureItem {
  name: string;
  market_value: number;
  percentage: number;
}

export interface SectorExposureResponse {
  as_of_date: string | null;
  portfolio_total: number;
  sector: ExposureItem[];
  message?: string;
}

export interface AccountSummary {
  id: number;
  account_name: string;
  account_type: string;
  currency: string;
  institution: string;
}

export interface SnapshotPoint {
  as_of_date: string;
  total_market_value: number;
  reporting_currency: string;
}

export interface KpiMetric {
  label: string;
  value: string;
  subValue: string;
  trend?: number[];
  positive?: boolean;
}

export interface InstitutionAllocation {
  name: string;
  value: number;
  percentage: number;
  connected: boolean;
}

export interface AccountAllocation {
  name: string;
  value: number;
  percentage: number;
}

export interface ScenarioComparison {
  title: string;
  actual: number;
  hypothetical: number;
}

export interface ImportRecord {
  source: string;
  timestamp: string;
  status: "Success" | "Failed" | "Pending";
}

export interface DashboardData {
  kpis: {
    totalValue: KpiMetric;
    dailyChange: KpiMetric;
    ytdReturn: KpiMetric;
    cashLiquidity: KpiMetric;
  };
  institutions: InstitutionAllocation[];
  accounts: AccountAllocation[];
  portfolioHistory: Record<string, Array<{ date: string; value: number }>>;
  sectorExposure: ExposureItem[];
  holdings: Array<{
    symbol: string;
    account: string;
    quantity: number;
    value: number;
    sector: string;
    currency: string;
    tags: string[];
  }>;
  scenario: ScenarioComparison;
  recentImports: ImportRecord[];
}

File: backend/app/services/portfolio.py
"""Portfolio-level aggregation service."""

from __future__ import annotations

from datetime import date

from sqlalchemy.orm import Session

from app.services._helpers import holding_market_value, holdings_for_date, latest_as_of_date


class PortfolioService:
    """Consolidated portfolio views backed by PostgreSQL holdings data."""

    def __init__(self, db: Session, reporting_currency: str = "CAD") -> None:
        self._db = db
        self._reporting_currency = reporting_currency

    def get_portfolio_summary(self, as_of: date | None = None) -> dict:
        holdings = holdings_for_date(self._db, as_of)
        if not holdings:
            return {
                "as_of_date": None,
                "reporting_currency": self._reporting_currency,
                "total_market_value": 0.0,
                "holding_count": 0,
                "account_count": 0,
                "institution_count": 0,
                "message": "No holdings found for the requested date.",
            }

        total = sum(holding_market_value(h) for h in holdings)
        accounts = {h.account_id for h in holdings}
        institutions = {h.account.institution_id for h in holdings}
        resolved_date = holdings[0].as_of_date

        return {
            "as_of_date": resolved_date.isoformat(),
            "reporting_currency": self._reporting_currency,
            "total_market_value": round(total, 2),
            "holding_count": len(holdings),
            "account_count": len(accounts),
            "institution_count": len(institutions),
        }

    def get_holdings(self, as_of: date | None = None) -> list[dict]:
        holdings = holdings_for_date(self._db, as_of)
        return [self._serialize_holding(h) for h in holdings]

    @staticmethod
    def _serialize_holding(holding) -> dict:
        instrument = holding.instrument
        account = holding.account
        return {
            "account_id": account.id,
            "account_name": account.account_name,
            "account_type": account.account_type,
            "institution": account.institution.name,
            "instrument_id": instrument.id,
            "symbol": instrument.symbol,
            "instrument_name": instrument.name,
            "asset_class": instrument.asset_class,
            "currency": instrument.currency,
            "quantity": holding.quantity,
            "market_price": holding.market_price,
            "market_value": round(holding_market_value(holding), 2),
            "as_of_date": holding.as_of_date.isoformat(),
        }
